backend/compact-connect/lambdas/python/disaster-recovery/handlers/rollback_license_upload.py

```diff
@@ -0,0 +1,1051 @@
+                            revision_id=reverted_privilege['revisionId'],
+                            action=reverted_privilege['action'],
+                        )
+                        for reverted_privilege in summary.get('privilegesReverted', [])
+                    ],
+                    updates_deleted=summary.get('updatesDeleted', []),
+                )
+                for summary in data.get('revertedProviderSummaries', [])
+            ],
+        )
+
+
+def rollback_license_upload(event: dict, context: LambdaContext):  # noqa: ARG001 unused-argument
+    """
+    Rollback invalid license uploads for a compact/jurisdiction/time window.
+
+    This function queries the licenseUploadDateGSI to find all affected records, validates
+    rollback eligibility, reverts records to their pre-upload state, and publishes events.
+    Results are written to S3 to avoid state management in the step function.
+
+    Input event structure:
+    {
+        'compact': 'aslp',
+        'jurisdiction': 'oh',
+        'startDateTime': '2024-01-01T00:00:00Z',
+        'endDateTime': '2024-01-01T23:59:59Z',
+        'rollbackReason': 'Invalid data uploaded',
+        'executionName': 'unique-execution-id',
+        'providersProcessed': 0,
+        'continueFromProviderId': None
+    }
+
+    Returns:
+    {
+        'rollbackStatus': 'IN_PROGRESS' | 'COMPLETE',
+        'providersProcessed': int,
+        'providersReverted': int,
+        'providersSkipped': int,
+        'providersFailed': int,
+        'continueFromProviderId': str | None,
+    }
+    """
+    execution_start_time = time.time()
+    max_execution_time = 12 * 60  # 12 minutes in seconds
+
+    # Extract and validate input parameters
+    compact = event['compact']
+    jurisdiction = event['jurisdiction']
+    start_datetime_str = event['startDateTime']
+    end_datetime_str = event['endDateTime']
+    rollback_reason = event['rollbackReason']
+    execution_name = event['executionName']
+    providers_processed = event.get('providersProcessed', 0)
+    continue_from_provider_id = event.get('continueFromProviderId')
+
+    # Parse and validate datetime parameters
+    try:
+        start_datetime = datetime.fromisoformat(start_datetime_str.replace('Z', '+00:00'))
+        end_datetime = datetime.fromisoformat(end_datetime_str.replace('Z', '+00:00'))
+    except ValueError as e:
+        logger.error(f'Invalid datetime format: {str(e)}')
+        return {
+            'rollbackStatus': 'FAILED',
+            'error': f'Invalid datetime format: {str(e)}',
+        }
+
+    # Validate time window
+    if start_datetime >= end_datetime:
+        logger.error('Start time must be before end time')
+        return {
+            'rollbackStatus': 'FAILED',
+            'error': 'Start time must be before end time',
+        }
+
+    time_window_seconds = (end_datetime - start_datetime).total_seconds()
+    if time_window_seconds > MAX_ROLLBACK_WINDOW_SECONDS:
```

Comment:
This is a lot of data loading and validation. Should we be using a data class and/or Schema here?